README
NAME
    Catalyst::Plugin::SubRequest - Make subrequests to actions in Catalyst

SYNOPSIS
        use Catalyst 'SubRequest';

        my $res_body = $c->subreq('/test/foo/bar', { template => 'magic.tt' });

        my $res_body = $c->subreq( {
           path            => '/test/foo/bar',
           body            => $body
        }, {
           template        => 'magic.tt'
        });

        # Get the full response object
        my $res = $c->subreq_res('/test/foo/bar', {
            template => 'mailz.tt'
        }, {
            param1   => 23
        });
        $c->log->warn( $res->content_type );

DESCRIPTION
    Make subrequests to actions in Catalyst. Uses the catalyst dispatcher,
    so it will work like an external url call. Methods are provided both to
    get the body of the response and the full response (Catalyst::Response)
    object.

METHODS
    subreq [path as string or hash ref], [stash as hash ref], [parameters as
    hash ref]
    subrequest
    sub_request
        Takes a full path to a path you'd like to dispatch to.

        If the path is passed as a hash ref then it can include body,
        action, match and path.

        An optional second argument as hashref can contain data to put into
        the stash of the subrequest.

        An optional third argument as hashref can contain data to pass as
        parameters to the subrequest.

        Returns the body of the response.

    subreq_res [path as string or hash ref], [stash as hash ref],
    [parameters as hash ref]
    subrequest_response
    sub_request_response
        Like "sub_request()", but returns a full Catalyst::Response object.

SEE ALSO
    Catalyst.

AUTHORS
    Marcus Ramberg, "mramberg@cpan.org"

    Tomas Doran (t0m) "bobtfish@bobtfish.net"

THANK YOU
    SRI, for writing the awesome Catalyst framework

COPYRIGHT
    Copyright (c) 2005 - 2011 the Catalyst::Plugin::SubRequest "AUTHORS" as
    listed above.

LICENSE
    This program is free software, you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Catalyst::Plugin::SubRequest

0.19  2012-07-19 23:50:58 BRT
        - Fixed non-propagation of parameters to subrequests

0.18  2012-06-20 08:58:54 BRT
        - Major refactor using Plack
        - Fix RT#77830 side-effect propagation to original request

0.17  2011-03-22 09:08:00 GMT
        - Added subrequest and subrequest_response names aliased to 
          sub_request and sub_request_response methods
        - Fix RT#66126 debug mode path display bug.

0.16  2010-01-28 21:09:04 GMT
        - Add subreq_res to get the full Catalyst::Response back
          from the sub request. RT#39486

0.15  2009-11-23 22:47:32 GMT
        - Be compatible with Catalyst 5.80014 by not writing to the
          request body directly.
        - Remove deprecation warnings by moving actions out of the test app.

0.14  2009-10-18 18:30:00 BST
        - Tweak copyright info to ease downstream packaging.

0.13  2008-01-06 22:33:00 CET
        - Update to work with new Stats package.

0.12  2007-09-12 13:08:00 CET
        - Switch to Module::Install
        - Fixed README (RT #16459)

0.11  2007-04-03 14:13:00
        - $c->req->params restored after sub_request

0.10  2005-12-08 13:15:00  
        - Updated test suite to work with new catalyst.

0.09  2005-10-28 01:07:00 
    	- localize body rather than output, which is an alias

0.08  2005-12-10 01:34:00 
	    - patch to handle NEXT trouble.

0.07  2005-08-01 12:08:00 
        - reset path.
        - allow passing params.

0.06  2005-06-08 20:27:00 
        - Flush params as well
        - allow stuff to be passed to the stash

0.05  2005-03-19 20:27:00 
        - Added a log message in debug mode.

0.04  2005-03-19 20:27:00
        - Updated forward to cat5 style
        - Using Catalyst::Dispatch::dispatch instead of forward.

0.03  2005-03-19 20:27:00
        - Fixed documentation.

0.02  2005-03-19 10:32:00
        - Ooops, missed MANIFEST

0.01  2005-03-05 22:00:00
        - first release


t/lib/TestApp.pm
package TestApp;

use Catalyst qw[SubRequest];

__PACKAGE__->config(
    name=>"subrequest test"
);

__PACKAGE__->setup();

    sub begin : Private {
        my ( $self, $c ) = @_;
        $c->res->body('1');
    }

    sub subtest : Global {
        my ( $self, $c ) = @_;
        my $subreq= $c->res->body().
                    $c->subreq('/normal/4');
        $c->res->body($subreq);
    }
  
    sub normal : Global {
        my ( $self, $c, $arg ) = @_;
        $c->res->body($c->res->body().$arg);
    }
    
    sub end : Private {
        my ( $self, $c ) = @_;
        $c->res->body($c->res->body().'3');
    }


1;


t/lib/TestApp/Controller/Root.pm
package TestApp::Controller::Root;
use strict;
use warnings;

use base qw/Catalyst::Controller/;

sub begin : Private {
    my ( $self, $c ) = @_;
    $c->res->body('1');
}

sub subtest : Global {
    my ( $self, $c ) = @_;
    my $subreq= $c->res->body().
                $c->subreq('/normal/4');
    $c->res->body($subreq);
}

sub normal : Global {
    my ( $self, $c, $arg ) = @_;
    $c->res->body($c->res->body().$arg);
}

sub subtest_params : Global {
    my ( $self, $c ) = @_;
    my $before = $c->req->params->{value};
    my $subreq = $c->subreq('/normal/2');
    my $after = $c->req->params->{value};
    $c->res->body($c->res->body().$after);
}

sub subtest_req : Global {
    my ( $self, $c ) = @_;
    my $subreq = $c->subreq('/normal/2');
    my $after = $c->req->uri->path;
    $c->res->body($after);
}

sub subtest_full_response : Global {
    my ( $self, $c ) = @_;
    my $subreq_res = $c->subreq_res('/typesetter');
    $c->res->body( $c->res->body() . $subreq_res->content_type );
}

sub subtest_with_params :Global {
  my($self, $c) = @_;
  $c->res->body(
    $c->subrequest('/plain_param', {},
      { content => 'foo' }));
}

sub plain_param :Global {
  my($self, $c) = @_;
  $c->res->body($c->req->params->{content});
}

sub typesetter : Global {
    my ( $self, $c, $arg ) = @_;
    $c->res->content_type( 'text/csv' );
    $c->res->body($c->res->body());
}

sub end : Private {
    my ( $self, $c ) = @_;
    $c->res->body($c->res->body().'3');
}

1;



lib/Catalyst/Plugin/SubRequest.pm
package Catalyst::Plugin::SubRequest;

use strict;
use warnings;
use Plack::Request;

our $VERSION = '0.19';

=head1 NAME

Catalyst::Plugin::SubRequest - Make subrequests to actions in Catalyst

=head1 SYNOPSIS

    use Catalyst 'SubRequest';

    my $res_body = $c->subreq('/test/foo/bar', { template => 'magic.tt' });

    my $res_body = $c->subreq( {
       path            => '/test/foo/bar',
       body            => $body
    }, {
       template        => 'magic.tt'
    });

    # Get the full response object
    my $res = $c->subreq_res('/test/foo/bar', {
        template => 'mailz.tt'
    }, {
        param1   => 23
    });
    $c->log->warn( $res->content_type );

=head1 DESCRIPTION

Make subrequests to actions in Catalyst. Uses the  catalyst
dispatcher, so it will work like an external url call.
Methods are provided both to get the body of the response and the full
response (L<Catalyst::Response>) object.

=head1 METHODS

=over 4

=item subreq [path as string or hash ref], [stash as hash ref], [parameters as hash ref]

=item subrequest

=item sub_request

Takes a full path to a path you'd like to dispatch to.

If the path is passed as a hash ref then it can include body, action,
match and path.

An optional second argument as hashref can contain data to put into the
stash of the subrequest.

An optional third argument as hashref can contain data to pass as
parameters to the subrequest.

Returns the body of the response.

=item subreq_res [path as string or hash ref], [stash as hash ref], [parameters as hash ref]

=item subrequest_response

=item sub_request_response

Like C<sub_request()>, but returns a full L<Catalyst::Response> object.

=back

=cut

*subreq              = \&sub_request;
*subrequest          = \&sub_request;
*subreq_res          = \&sub_request_response;
*subrequest_response = \&sub_request_response;

sub sub_request {
  return shift->sub_request_response(@_)->body;
}

sub sub_request_response {
  my ( $c, $path, $stash, $params ) = @_;
  $stash ||= {};
  my $env = $c->request->env;
  my $req = Plack::Request->new($env);
  my $uri = $req->uri;
  $uri->query_form( $params || {} );
  local $env->{QUERY_STRING} = $uri->query || '';
  local $env->{PATH_INFO}    = $path;
  local $env->{REQUEST_URI}  = $env->{SCRIPT_NAME} . $path;
  $env->{REQUEST_URI} =~ s|//|/|g;
  my $class = ref($c) || $c;

  $c->stats->profile(
    begin   => 'subrequest: ' . $path,
    comment => '',
  ) if ( $c->debug );

  # need this so that
  my $writer = Catalyst::Plugin::SubRequest::Writer->new;
  my $response_cb = sub { $writer };
  my $i_ctx = $class->prepare( env => $env, response_cb => $response_cb );
  $i_ctx->stash($stash);
  $i_ctx->dispatch;
  $i_ctx->finalize;
  $c->stats->profile( end => 'subrequest: ' . $path ) if $c->debug;

  $i_ctx->response->body($writer->body);

  return $i_ctx->response;
}

package Catalyst::Plugin::SubRequest::Writer;
use Moose;
has body => (
  isa     => 'Str',
  is      => 'ro',
  traits  => ['String'],
  default => '',
  handles => { write => 'append' }
);
has _is_closed => ( isa => 'Bool', is => 'rw', default => 0 );
sub close { shift->_is_closed(1) }

around write => sub {
  my $super = shift;
  my $self = shift;
  return if $self->_is_closed;
  $self->$super(@_);
};

=head1 SEE ALSO

L<Catalyst>.

=head1 AUTHORS

Marcus Ramberg, C<mramberg@cpan.org>

Tomas Doran (t0m) C<< bobtfish@bobtfish.net >>

=head1 MAINTAINERS

Eden Cardim (edenc) C<eden@insoli.de>

=head1 THANK YOU

SRI, for writing the awesome Catalyst framework

MIYAGAWA, for writing the awesome Plack toolkit

=head1 COPYRIGHT

Copyright (c) 2005 - 2011
the Catalyst::Plugin::SubRequest L</AUTHORS>
as listed above.

=head1 LICENSE

This program is free software, you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut

1;
